### DISCUSSION TITLE
About 3GPP R16 AVPs Support

### DISCUSSION THREAD

**Discussion by thakanoduncu** [freeDiameter/freeDiameter#58] [Q&A] [answered]
Hello. I'm a software engineer working for a telecommunications company currently developing 4G and 5G solutions.

First of all, I want to thank you for providing this great open-source library that makes lots of contributions to our works. In order to develop services requiring Diameter protocol implementation, we've been using freeDiameter library for a few years.

Recently, we've decided to upgrade our 5G solutions 3GPP version from R15 to R16 and started to implement R16 requirements including diameter based services. During development we have realized that some AVPs introduced in R16 are missing in related dictionaries included in freeDiameter. For instance, SMSF related sub AVPs of Serving-Node (2401) and Additional-Serving-Node  AVP (2406) are not listed in 3GPP dictionary ([contrib/dict_json/dictionary_dcca_3gpp.json](https://github.com/freeDiameter/freeDiameter/blob/a3dd2700369e20ff4edaf75042a0721f86a8081b/contrib/dict_json/dictionary_dcca_3gpp.json)).

Missing AVPs are listed in 3GPP 29.337 v16.1 as seen on images below.
![image](https://github.com/freeDiameter/freeDiameter/assets/10131051/95eeb3cf-fd84-41ee-bb36-a30fb8412f2d)

![image](https://github.com/freeDiameter/freeDiameter/assets/10131051/f6a3650e-61de-461f-a0cf-fbc8dc486f7a)

So, I'd like to ask if there is a scheduled update to library in order to support some R16 features and related AVPs soon. If not, I'd like to have your opinions about what kind of solutions I may use to overcome this issue? Any assistance would be greatly appreciated.

Kind Regards,
Tolga Hakan Oduncu

**Comment by sdecugis** [comment]
Hello,

You can propose a patch to add the missing definitions, and i will check to
merge it. It s probably the easier way at the moment.

Best regards,
Sebastien

**Reply by lukem** [reply]
Hi Tolga,

There's also the separate extensions/dict_dcca_3gpp module which can be used instead of the JSON module.

The extension uses the source file `extensions/dict_dcca_3gpp/add_avps.c` which is generated from `extensions/dict_dcca_3gpp/dict_dcca_3gpp.csv` using `contrib/tools/csv_to_fd` using:

```
cd extensions/dict_dcca_3gpp
../../contrib/tools/csv_to_fd -p fdc dict_dcca_3gpp.csv > add_avps.c
```

The file `dict_dcca_3gpp.csv` is managed by copypasta relevant tables from the various specs into the CSV, with specific annotations and comments. The top of `csv_to_fd` has comments explaining the format of the csv.

I implemented this 4 years ago in early 2020. At the time, I updated the dictionary to the Release 15 files at the time.


Your discussion reminds me that at some point someone (probably me if I get time) should review of the changes between the versions currently described in `dict_dcca_3gpp.csv` and the latest Word documents, for Release 16, Release 17, or even the unfrozen Release 18.

One issue I recall is that we don't have an easy way to deal with AVPs being renamed, in a backwards compatible manner for downstream consumers. I'm not sure we need to solve that right now. For example, at some point someone "helpfully" (and incorrectly, IMHO), renamed various parts of the specs from "IPv4" and "IPv6" to "Ipv4" and "Ipv6", which had the effect of renaming AVPs... See commit d49b72e in the repo. I think that this may have been reverted more recently - I did mention it offhand to a 3GPP technical officer and it may now be fixed, but I'll have to check again.


(On using the `csv_to_fd` tool today, using python 3, I've observed that it has nondeterministic ordering for certain output, probably because it's using a set() or dict() without ensuring ordering, and with python3 this because less deterministic. I'll look at fixing that)

**Reply by thakanoduncu** [reply]
Hi again,
Thank you for your quick responses.

@lukem What I understand from your first sentences is that I can update/modify those dictionary files by myself and build the library for a custom version. Is that correct?
Besides, your updates to repository would be a better permanent solution. I clearly understand your hesitations about backward compatibility. In order to overcome any possible issue, maybe you should release different builds from different branches created for R16, R17, etc. if possible.

**Comment by lukem** [comment, marked as answer]
You definitely can update the dictionaries yourself, and use locally.
If the changes were provided as a pull request, we can review and merge in the baseline too.

I don't think we'd want to make different builds from different branches. If necessary, we could create variations of dict_dcca_3gpp that were specific to a 3GPP Release, but I'm not sure we'd need to do that either. The backwards incompatible rename would only affect downstream consumers of freeDiameter's dict_dcca_3gpp module; it doesn't affect the "on wire" encoding of the Diameter AVPs. We could continue to document the changes in release notes and people could adapt as required.
